### PR TITLE
[Estuary] Video Version skin redesign + cleanup

### PR DESCRIPTION
**Pull request by jjd-uk**
## Description

Redesign of the Estuary changes for the new Video Version feature, plus some some minor cleanup.

## Screenshots (if appropriate):

**Dialog Video Version - Play Mode before**

image

**Dialog Video Version  - Play mode after**

Introduced buttons to select whether Versions or Extras list is shown

image

**Dialog Video Version - Play Mode before - Above 3 extras**

Only 3 items in list can be shown

image

**Dialog Video Version  - Play mode after - Above 3 extras**

Increased list space allows more than 3 items to be shown, and if number of items in the list exceeds the list space then a scrollbar is shown as is done for other lists. In Lawrence of Arabia example there are 9 extras.

image

**Dialog Video Version - Play Mode Before - No extras**

image

**Dialog Video Version - Play Mode After - No extras**

Extras button is not shown and Versions button is disabled

image

**Dialog Video Version - Manage Mode Before**

image

**Dialog Video Version - Manage Mode After - Versions**

Add extras button hidden

image

**Dialog Video Version - Manage Mode After - Extras**

Add version and Set default buttons hidden

image

**Dialog Video Info - Before**

Versions button was at beginning.

image

**Dialog Video Info - After**

Versions button moved to be alongside Sets button as all manage style buttons are in towards end of list.

image

### CODE REVIEW

**Comment by enen92** [comment]
Nice one 👍

**Comment by jjd-uk** [comment]
In an effort to improve the user experience I've given Play and Manage modes slightly different looks while still based on common code. I've introduced more conditional so anything not appropiate to the mode of oppperation are hidden, so this needs testing to make sure I've covered all use cases.

@xodidox while what you did was great for a first effort at skinning, I felt there could be improvements.

**Comment by ksooo** [comment]
I like that, thanks @jjd-uk.

**Comment by jjd-uk** [comment]
For everyones info, I based the code implemtatation to be similar in style to that done for Dialog Select where **DialogSelect.xml** uses includes for different layouts and **Includes_DialogSelect.xml** hosts some of the includes.

**Comment by garbear** [comment]
Nice improvements!

**Comment by xodidox** [comment]
@jjd-uk , seems we took different design philosophies.

I like a stable UI, for buttons that not work in current selection, just disable them, thus I can always expect the location of a button. For the lists, I followed the "Add video source" dialog to present items with a variable count (obviously I missed the scrollbar here, but you can scroll in fact). For the "Video"/"Extras", I followed the "Set content" dialog.

With my design, you just need press "up"/"down" to choose item ("left" to quickly switch between "Versions" and "Extras"), and "right" to do action. The left part is only for artwork, the UI is clean.

With your design, one more button pressing needed to switch "Versions" and "Extras", and mixed buttons and artwork in left side is not my taste.

I'm not against any UI improvement and understand everyone has their own taste. Wish we can do a pool in forum (or other places) to decide what to use.

**Comment by HitcherUK** [comment]
> With my design, you just need press "up"/"down" to choose item ("left" to quickly switch between "Versions" and "Extras")

While that is neat this type of control isn't used anywhere else in Estuary and as such it's not intuitive.

**Comment by jjd-uk** [comment]
If in Play mode why would I want to see action buttons that can't be used?  It's unnecessary clutter IMHO.

If in Play mode and have no Extras why would I want the list area for Extras shown? Doing away with dual lists shown at the same time allows more space for each list, which is particularly useful as I've shown for movies that have lots of extras.

**Comment by da-anda** [comment]
I personally would go as far as splitting extras from versions, as they to me are quite different things. Regular play action would then only provide an option to select which version to play, while the extras can either be played via context menu or via a dedicated button in the videoInfoDialog. At least that would be my vision of this feature in the long term. For a first implementation having things mixed is ok.
Long term we might even consider having the behvior configurable, just like what we do for BluRays (play main movie, simplified menu, full bluray menu). So for this feature it could be
- play default version
- choose version
- choose version or extra

That way it could be tuned to personal preferences. But for now, having it combined should be fine (unless somebody has the time to split it up before we go into beta).

File: addons/skin.estuary/xml/Includes_DialogVideoVersion.xml [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+			<height>180</height>
+			<orientation>vertical</orientation>
+			<itemgap>dialogbuttons_itemgap</itemgap>
+			<onright condition="ControlGroup(200).HasFocus(201)">50</onright>
+			<onright condition="ControlGroup(200).HasFocus(202)">51</onright>
+			<include content="DefaultDialogButton">
+				<param name="id" value="201" />
+				<param name="visible">Window.IsActive(videoversion)| Window.IsActive(videoversionplay) + !String.IsEmpty(Container(50).ListItem.Label)</param>
+				<param name="enable">!String.IsEmpty(Container(51).ListItem.Label) | Window.IsActive(videoversion)</param>
+				<param name="label" value="Versions" />
```
Comment (jjd-uk) [inline addons/skin.estuary/xml/Includes_DialogVideoVersion.xml:90]: To be corrected - temporary hack still has not been removed, I need to set correct localised label of $LOCALIZE[40000]

File: addons/skin.estuary/xml/Includes_DialogVideoVersion.xml [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+			</include>
+			<include content="DefaultDialogButton">
+				<param name="id" value="202" />
+				<param name="visible">Window.IsActive(videoversion) | Window.IsActive(videoversionplay) + !String.IsEmpty(Container(51).ListItem.Label)</param> 
+				<param name="label" value="Extras" />
```
Comment (jjd-uk) [inline addons/skin.estuary/xml/Includes_DialogVideoVersion.xml:95]: To be corrected - temporary hack still has not been removed, I need to set correct localised label of $LOCALIZE[31615]

**Comment by jjd-uk** [comment]
@jurialmunkey stated it far better than me why I've taken the approach I have with the PR, so in summary:

Single List shown at any one time.
Buttons to switch the list displayed.
Buttons not appropiate to the selected list are hidden, for example the Rename/Set Default/Choose Art buttons can't be used for Extras so these are hidden when the Extras list is selected.

As for the design, it was based on the layout of the addon settings dialog where the categories are on the left switches the setting list displayed, then the settings list is in the middle, and buttons on the right.

image

**Comment by jjd-uk** [comment]
PR updated to remove a couple of hacks I missed that were still there from testing. In addition the code has been altered slightly after some code suggestions from @hitcher

For now I've removed the condition that disabled the Version button when it's the only button displayed where this no Extras thand thus the Extras button is hidden (in play mode only)

**Comment by jjd-uk** [comment]
Main change in update is the selected Version / Extra button retains the focus colour at at 50% when you move to the list or the buttons on the right to keep that visual clue.

image

image

Beyond the scope of this PR, but we really should do something about the selected item colour, white text for selected in a grey text list does not standout very well.

**Comment by xodidox** [comment]
I'd avoid disputing the expertise of Kodi skin development gurus like @jurialmunkey as it wouldn't be a wise decision. 😅

I'll reserve my design idea if most skinners think this is a better design. One issue here though, the "Choose art" button should be enabled for video extras too, as the only artwork for video extras by default is thumbnail which is extracted during scanning (if allowed by user). Enable this button will let user choose different artworks.

> On this topic, I noticed a broader problem here: currently (at least from using the most recently nightly) it appears you have to first click on the item to change which item the buttons are editing. Simply focusing the item then pressing right does not change which item is being acted upon. This makes it even more confusing as to which item you're editing.

I'll see how I can improve this part.

**Review by Hitcher — APPROVED** [review]
All good, thanks.

**Comment by jjd-uk** [comment]
@xodidox if we go ahead and merge this, you can keep your design by using your version of DialogVideoVersion.xml the only thing missing would be the new button you created VideoVersionDialogButton as I didn't see any need for that when we had other button types that could be reused.

If you wanted to keep your look exactly as before then you could simply drop in your DialogVideoVersion.xml and change the grouplist id 101 buttons to:

```
				<include content="DefaultDialogButton">
					<param name="id" value="27" />
					<param name="label" value="$LOCALIZE[13511]" />
					<param name="width">280</param>
					<param name="height">120</param>
					<param name="font">font13</param>
				</include>
```

**Comment by jjd-uk** [comment]
> @jjd-uk please add the "Choose art" button for extras before merge.

Sure I can do that. Before changing anything I was thinking perhaps:

- Play (applies to Versions & Extras)
- Add Version (applied to Versions only)
- Rename (applied to Versions only)
- Set default (applied to Versions only)
- Choose Art (applied to Versions only)
- Add Extras (applies to Extras only)
- Remove (applies to Versions & Extras)

So all the Versions only buttons are grouped together.

**Comment by xodidox** [comment]
> > @jjd-uk please add the "Choose art" button for extras before merge.
> 
> Sure I can do that. Before changing anything I was thinking perhaps:
> 
> * **Play** (applies to Versions & Extras)
> * **Add Version** (applied to Versions only)
> * **Rename** (applied to Versions only)
> * **Set default** (applied to Versions only)
> * **Choose Art** (applied to Versions only)
> * **Add Extras** (applies to Extras only)
> * **Remove** (applies to Versions & Extras)
> 
> So all the Versions only buttons are grouped together.

"Rename" also applies to extras as the name is generated automatically based on file name, user may want to change that.

If only apply "Add extras" to extras, for a video without extras, there is no button to add the first extras.

**Comment by jjd-uk** [comment]
> "Rename" also applies to extras as the name is generated automatically based on file name, user may want to change that.

That's not obvious, as when you select rename it's all the included version names that appears, and I don't see any obvious way to rename an Extra? This is what I see when I select an Extra and go to Rename

image
 
> If only apply "Add extras" to extras, for a video without extras, there is no button to add the first extras.

It's there in Manage mode.

**Comment by jjd-uk** [comment]
> > That's not obvious, as when you select rename it's all the included version names that appears, and I don't see any obvious way to rename an Extra? This is what I see when I select an Extra and go to Rename
> 
> As the extras names are generated automatically, so I didn't include them in the version list, but if user create some names (for example, "behind the scenes"), they can choose that from the end of the list. They can also use the "new version" button to create a new name for it.

Why do you even see these at all then? as it's an extra I want to rename I'm not expecting to see a long list of version type names.

**Comment by jjd-uk** [comment]
In addition the title of dialog is **Select movie version** so no mention of Extras, so I was mislead in thinking it wasn't for Extras.

**Comment by xodidox** [comment]
> In addition the title of dialog is **Select movie version** so no mention of Extras, so I was mislead in thinking it wasn't for Extras.

This is a bug.

**Comment by jjd-uk** [comment]
@xodidox Thanks for your continued efforts on this, it's much appreciated.

**Comment by jjd-uk** [comment]
@xodidox updated

Manage mode for Versions

image

Manage mode for Extras

image

**Comment by jurialmunkey** [comment]
@jjd-uk - I believe choose art is meant to be available for extras too, not just versions.

**Comment by jurialmunkey** [comment]
> 
> If only apply "Add extras" to extras, for a video without extras, there is no button to add the first extras.

@xodidox that shouldn't be an issue. As long as "extras" button is available on the left, you should be able to focus it and then when you press right kodi will leapfrog the blank list to move directly to the right side buttons where you add extras. 

This layout will also make it very simple for the skin to allow for the dialog to be split into separate versions and extras dialogs launched from separate buttons. All it will take is the skin having a button which sets a window property before doing a `sendclick(14)` and then skinning the dialog to only show either versions or extras depending on the property value (then clear onunload so that context menu entry will still open with both visible)

**Comment by jjd-uk** [comment]
When I looked at it I got the impression that Choose Art was for the video versions only. Will recheck.

**Comment by JDRIVO** [comment]
Personally, I wouldn't add a play button to the management section as it has no relevance in the context of version management.

**Comment by JDRIVO** [comment]
I'd also reduce the padding of the artwork to make it appear more connected if you catch my drift.

manage

**Comment by HitcherUK** [comment]
The play button could be dropped altogether if simply highlighting an item then selects it for editing. The click to play would be consistent across both dialogs.

**Comment by jjd-uk** [comment]
The click to select is what is done on all other dialogs so I don't think that should be changed.

**Comment by da-anda** [comment]
> Personally, I wouldn't add a play button to the management section as it has no relevance in the context of version management.

I agree - a manage dialog shouldn't have/need a play button - but I have no hard feelings about it

**Comment by jjd-uk** [comment]
Tweaked the right hand side button layout again in Manage mode. Somehow I missed **Choose art** was also applicable to Extras as well.

Position 2 button is alaways the add button - flips between Version and Extras dependant on what's selected on the left.
Positions 3 / 4 / 5 are always - Choose art, Rename, Remove
Position 6 is Set default which only applies to Versions

So everything is in a consistent place.

image

image

**Comment by jjd-uk** [comment]
I will give the weekend for any objections then will merge on Monday if none received.

@JDRIVO I understand your point and it's a valid one. I would like to see more feedback from people using the versions feature as there's several labels that may need to be looked at.

**Pull request merged**